gauntlet/packages/gauntlet-solana/src/commands/internal/solana.ts

```diff
@@ -86,24 +139,204 @@ export default abstract class SolanaCommand extends WriteCommand<TransactionResp
+
     const tx = makeTx(
-      rawTxs,
       {
-        blockhash,
-        lastValidBlockHeight,
-        feePayer: this.wallet.publicKey,
+        instructions: rawTxs,
+        recentBlockhash: currentBlockhash.blockhash,
+        payerKey: this.wallet.publicKey,
       },
       overrides,
     )
     if (extraSigners) {
-      tx.sign(...extraSigners)
+      tx.sign(extraSigners)
     }
-    const signedTx = await this.wallet.signTransaction(tx)
+    const signedTx = await this.wallet.signVersionedTransaction(tx)
     logger.loading('Sending tx...')
-    return await sendAndConfirmRawTransaction(this.provider.connection, signedTx.serialize())
+
+    const rawTransaction = signedTx.serialize()
+
+    const txid = await this.provider.connection.sendRawTransaction(rawTransaction, {
+      skipPreflight: true,
+    })
+
+    console.log('Started awaiting confirmation for', txid, 'size:', rawTransaction.length)
+    const startTime = getUnixTs()
+    let timeout = 60_000
+    let done = false
+    let retryAttempts = 0
+    const retrySleep = 2000
+    const maxRetries = 30
+    ;(async () => {
+      while (!done && getUnixTs() - startTime < timeout / 1000) {
+        await sleep(retrySleep)
+        // console.log(new Date().toUTCString(), ' sending tx ', txid);
+        this.provider.connection.sendRawTransaction(rawTransaction, {
+          skipPreflight: true,
+        })
+        if (retryAttempts <= maxRetries) {
+          retryAttempts = retryAttempts++
+        } else {
+          break
+        }
+      }
+    })()
+
+    try {
+      await this.awaitTransactionSignatureConfirmation(txid, timeout, CONFIRM_LEVEL, currentBlockhash)
+    } catch (err: any) {
+      if (err.timeout) {
+        throw new TimeoutError({ txid })
+      }
+      let simulateResult: SimulatedTransactionResponse | null = null
+      try {
+        simulateResult = (await this.provider.connection.simulateTransaction(tx)).value
+      } catch (e) {
+        console.warn('Simulate transaction failed')
+      }
+
+      if (simulateResult && simulateResult.err) {
+        if (simulateResult.logs) {
+          for (let i = simulateResult.logs.length - 1; i >= 0; --i) {
+            const line = simulateResult.logs[i]
+            if (line.startsWith('Program log: ')) {
+              throw new SolanaError({
+                message: 'Transaction failed: ' + line.slice('Program log: '.length),
+                txid,
+              })
+            }
+          }
+        }
+        throw new SolanaError({
+          message: JSON.stringify(simulateResult.err),
+          txid,
+        })
+      }
+      throw new SolanaError({ message: 'Transaction failed', txid })
+    } finally {
+      done = true
+    }
+    return txid
+
+    // TODO: skipPreflight: true since we previously simulated it?
+    // return await sendAndConfirmRawTransaction(this.provider.connection, rawTransaction) // TODO: { maxRetries: 5 }
+  }
+
+  async awaitTransactionSignatureConfirmation(
```

Comment:
This whole fn is almost impossible to understand. I know it comes from Mango, but some refactor or at least comments I feel is necessary. The next one who could come here to update something will appreciate it